### PR TITLE
[PROD-779] Add total staked to farm page

### PR DESCRIPTION
**Pull request by leqiang-deltafi**
Replace last deltafi price with total staked value
Move stake values calculation logic to farm component, and farm component passes corresponding values to farmCard
Screen Shot 1401-03-31 at 02 04 39

### CODE REVIEW

File: src/utils/utils.ts [hunk 1]
```diff
@@ -39,7 +39,7 @@ export function getTokenTvl(tokenConfig: TokenConfig, amount: BN, price: BigNumb
   return anchorBnToBn(tokenConfig, amount).multipliedBy(price);
 }
 
-export function getUserTokenTvl(tvl: BigNumber, share: BN, supply: BN) {
+export function getStakedTokenTvl(tvl: BigNumber, share: BN, supply: BN) {
```
[the remark this answers is not in the document]
Comment (ypeng-deltafi) [reply src/utils/utils.ts:42]: can you keep it as getUserTokenTvl or getTokenShareTvl

File: src/views/Dashboard/Home.tsx [hunk 2]
```diff
@@ -37,6 +38,7 @@ import { calculateWithdrawalFromShares } from "lib/calc";
 import { formatCurrencyAmount } from "utils/utils";
 import { CircularProgress } from "@material-ui/core";
 import { dashboardViewActions } from "states/views/dashboardView";
+import { calculateFarmPoolsData } from "views/Farm/utils";
```
Comment (ypeng-deltafi) [inline src/views/Dashboard/Home.tsx:41]: Data is too general. can you give it a more specific name?

Comment (ypeng-deltafi) [reply src/views/Dashboard/Home.tsx:41]: is it farm pool tvl?

File: src/views/Farm/Home.tsx [hunk 3]
```diff
@@ -237,60 +284,58 @@ const Home: React.FC = (props): ReactElement => {
...
-                            key={farm.farmInfo}
-                            poolConfig={poolConfig}
-                            farmInfoAddress={farm.farmInfo}
-                          />
-                        </Grid>
-                      )),
-                  )}
-                </Grid>
-              )}
+              <Grid container className={classes.poolCardContainer} justifyContent="center">
```
Comment (ypeng-deltafi) [inline src/views/Farm/Home.tsx:287]: this part looks duplicated. shall we make it a function?

File: src/views/Dashboard/Home.tsx (hunk 2 above)
Comment (leqiang-deltafi) [reply src/views/Dashboard/Home.tsx:41]: not just farm pool tvl, includes each pool's total tvl, user tvl and apr

File: src/utils/utils.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (leqiang-deltafi) [reply src/utils/utils.ts]: updated

File: src/views/Farm/Home.tsx [hunk 4]
```diff
@@ -165,28 +172,68 @@ const Home: React.FC = (props): ReactElement => {
...
+    ],
+  );
+
+  // filter out active and inactive farms
+  const { activeFarms, inactiveFarms } = useMemo(() => {
+    const activeFarms: FarmInfoData[] = [];
+    const inactiveFarms: FarmInfoData[] = [];
+
+    fullFarmPoolsData.forEach((farmInfoData) => {
+      const farmConfig = farmKeyToFarmInfo[farmInfoData.farmInfoAddress]?.farmConfig;
```
Comment (ypeng-deltafi) [inline src/views/Farm/Home.tsx:207]: why not just put the farmConfig to calculateFarmPoolsData output as well?

File: src/views/Farm/Home.tsx [hunk 5]
```diff
@@ -237,60 +284,58 @@ const Home: React.FC = (props): ReactElement => {
...
-                      )),
-                  )}
-                </Grid>
-              )}
+              <Grid container className={classes.poolCardContainer} justifyContent="center">
+                {activeFarms.map(
+                  ({ farmInfoAddress, totalStaked, userStaked, apr, poolConfig }, idx) => (
+                    <Grid item key={idx} xl={2} lg={3} md={4} sm={6}>
+                      <FarmCard
+                        color={
```
Comment (ypeng-deltafi) [inline src/views/Farm/Home.tsx:292]: do you think we can pass the `index`  to FarmCard and let it handle the color?

File: src/views/Farm/Home.tsx [hunk 6]
```diff
@@ -237,60 +284,58 @@ const Home: React.FC = (props): ReactElement => {
...
+                            ? "greenYellow"
+                            : idx % 4 === 1
+                            ? "lime"
+                            : idx % 4 === 2
+                            ? "indigo"
+                            : "dodgerBlue"
+                        }
+                        key={farmInfoAddress}
+                        poolConfig={poolConfig}
+                        farmInfoAddress={farmInfoAddress}
```
Comment (ypeng-deltafi) [inline src/views/Farm/Home.tsx:330]: do we need both key and farmInfoAddress? they have same value

File: src/views/Farm/utils.ts [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+  poolConfigs.forEach((poolConfig) => {
+    const swapInfo = swapKeyToSwapInfo[poolConfig.swapInfo];
+    poolConfig?.farmInfoList?.forEach(({ farmInfo: farmInfoAddress }) => {
+      const farmInfo = farmKeyToFarmInfo[farmInfoAddress];
+      const farmUser = farmPoolKeyToFarmUser[farmInfoAddress];
+      const basePrice = symbolToPythPriceData[poolConfig.base];
+      const quotePrice = symbolToPythPriceData[poolConfig.quote];
+
+      if (
```
Comment (ypeng-deltafi) [inline src/views/Farm/utils.ts:36]: can you reverse the if-else to reduce the nested level? 
```
if (!condition) {
 return {
  special value
 };
}

normal case
```

Comment (ypeng-deltafi) [reply src/views/Farm/utils.ts:36]: otherwise, people need to scroll down a lot to see the else.

Comment (leqiang-deltafi) [reply src/views/Farm/utils.ts:36]: I see, updated

File: src/views/Farm/Home.tsx (hunk 3 above)
Comment (leqiang-deltafi) [reply src/views/Farm/Home.tsx]: added a local util function here

File: src/views/Dashboard/Home.tsx [hunk 8]
```diff
@@ -365,28 +389,30 @@ const Home: React.FC = (props) => {
...
-                          />
-                        </Grid>
-                      )),
-                    )}
+                  {userFarmPoolsData.map(
+                    ({ farmInfoAddress, totalStaked, userStaked, apr, poolConfig }, idx) => (
+                      <Grid item key={idx} xl={2} lg={3} md={4} sm={6}>
+                        <FarmCard
+                          color={
+                            idx % 4 === 0
```
Comment (ypeng-deltafi) [inline src/views/Dashboard/Home.tsx:397]: I think we pass a list of `colors` and index to FarmCard at least

File: src/views/Farm/Home.tsx (hunk 4 above)
Comment (ypeng-deltafi) [reply src/views/Farm/Home.tsx:207]: I meant if you add farmConfig to fullFarmPoolsStakeInfo, you don't need to use farmKeyToFarmInfo here

File: src/views/Farm/Home.tsx (hunk 6 above)
Comment (leqiang-deltafi) [reply src/views/Farm/Home.tsx]: The key field was there and was duplicated with farmInfoAddress. Looks like the key is useless and i just removed it

**Pull request merged**